doc/developers/cython.rst

```diff
@@ -0,0 +1,178 @@
+* No concept similar to interfaces or traits exist in Cython
+* Currently it is impossible to inherit and override a method which has a fused type in its signature
+* Structs cannot be created using constructor syntax with `nogil` `issue and workaround <https://github.com/cython/cython/issues/1642>`_
+
+Performances improvements will come with `Cython 3.0 which is to be released soon in beta apparently <https://github.com/cython/cython/issues/4022#issuecomment-1340918590>`_
+
+Basics
+^^^^^^
+
+* ``nogil`` declaration: https://github.com/cython/cython/issues/2798#issuecomment-459971828
+
+Tips for developping Cython code in scikit-learn
+------------------------------------------------
+
+Tips to ease development
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+* If you intend to use OpenMP: On MacOS, system's distribution of ``clang`` does not implement OpenMP.
+  You can install the ``compilers`` package available on ``conda-forge`` which comes with implementations of OpenMP.
+
+* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:
+
+  .. code-block:: bash
+
+         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1
+
+* Start from scratch in a notebook to understand.
+* To debug C code (e.g. a segfault), do use ``gdb`` with:
+
+  .. code-block:: bash
+
+         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py
+
+* To have access to some value in place to debug in ``cdef (nogil)`` context, use:
+
+  .. code-block:: cython
+
+         with gil:
+             print(state_to_print)
+
+* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.
+    ``print(f"{test_val=}")``
+
+* scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and we need to simplified
+  and unify that across the codebase. For now, make sure you understand which concrete types are used ultimately.
+
+* It is helpful to use ``gdb`` to debug. In order to do so, one must use a Python interpreter built with debug support
+  (debug symbols and proper optimization). To create a new conda environment with a source-built interpreter:
+
+  .. code-block:: bash
+
+         git clone https://github.com/python/cpython.git
+         conda create -n debug-scikit-dev
+         conda activate debug-scikit-dev
+         cd cpython
+         mkdir debug
+         cd debug
+         ../configure --prefix=$CONDA_PREFIX --with-pydebug
+         make EXTRA_CFLAGS='-DPy_DEBUG' -j<num_cores>
+         make install
+
+* You may need to restart conda after building/installing (I had to, but not sure why).
+
+
+  * You might find this alias handy:
+    .. code-block::
+
+         alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"
+
+  * which you can use with:
+
+    .. code-block::
+
+         cythonX --annotate source.pyx
```

Comment:
I think a separate point about ``--annotate`` is worth. You can explain what it does where to find the annotated code (in the ``.html``) and how to interpret it: the more yellow is the line the more interaction with the Python interpreter it generates; we want to avoid such interactions as much as possible in parts of the code that are called many times, e.g. inside loops.

Reply:
What you propose is valuable, indeed. Yet, I think it might be stated into Cython's documentation directly.

What do you think?

Reply:
There's definitely some examples of annotation in the Cython documentation (https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes). Although I don't believe for a second we have everything right and in its best place so improvements are welcome.

Personally, I don't think it hurts if there's a bit of duplication of advice.